File: ai-agent-search-optimization/scripts/audit_site.py
#!/usr/bin/env python3
"""Dependency-free AI search visibility audit for a single website URL."""

from __future__ import annotations

import argparse
import html
import json
import re
import sys
import urllib.error
import urllib.parse
import urllib.request
import urllib.robotparser
from html.parser import HTMLParser
from typing import Any


DEFAULT_AGENTS = [
    "Googlebot",
    "OAI-SearchBot",
    "GPTBot",
    "ChatGPT-User",
    "PerplexityBot",
    "Perplexity-User",
    "ClaudeBot",
    "CCBot",
]


class FetchResult(dict):
    pass


class PageParser(HTMLParser):
    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.title = ""
        self._in_title = False
        self._skip_depth = 0
        self.meta: list[dict[str, str]] = []
        self.links: list[dict[str, str]] = []
        self.headings: dict[str, list[str]] = {"h1": [], "h2": []}
        self.json_ld: list[str] = []
        self._in_json_ld = False
        self._json_ld_parts: list[str] = []
        self.text_parts: list[str] = []

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        tag = tag.lower()
        attrs_dict = {k.lower(): (v or "") for k, v in attrs}
        if tag in {"script", "style", "noscript"}:
            self._skip_depth += 1
        if tag == "title":
            self._in_title = True
        if tag == "meta":
            self.meta.append(attrs_dict)
        if tag == "link":
            self.links.append(attrs_dict)
        if tag == "script" and attrs_dict.get("type", "").lower() == "application/ld+json":
            self._in_json_ld = True
            self._json_ld_parts = []
        if tag in {"h1", "h2"}:
            self.text_parts.append(f"\n<{tag}>")

    def handle_endtag(self, tag: str) -> None:
        tag = tag.lower()
        if tag in {"script", "style", "noscript"} and self._skip_depth:
            self._skip_depth -= 1
        if tag == "title":
            self._in_title = False
        if tag == "script" and self._in_json_ld:
            self.json_ld.append("".join(self._json_ld_parts).strip())
            self._in_json_ld = False
            self._json_ld_parts = []

    def handle_data(self, data: str) -> None:
        text = " ".join(data.split())
        if not text:
            return
        if self._in_title:
            self.title += text
        elif self._in_json_ld:
            self._json_ld_parts.append(data)
        elif not self._skip_depth:
            self.text_parts.append(text)

    def close(self) -> None:
        super().close()
        text = " ".join(self.text_parts)
        for level in ("h1", "h2"):
            pattern = re.compile(rf"<{level}>\s*([^<]+)", re.IGNORECASE)
            self.headings[level] = [m.group(1).strip() for m in pattern.finditer(text)]


def normalize_url(url: str) -> str:
    if not re.match(r"^https?://", url, re.I):
        url = "https://" + url
    parsed = urllib.parse.urlparse(url)
    path = parsed.path or "/"
    return urllib.parse.urlunparse((parsed.scheme, parsed.netloc, path, "", parsed.query, ""))


def origin(url: str) -> str:
    parsed = urllib.parse.urlparse(url)
    return f"{parsed.scheme}://{parsed.netloc}"


def fetch(url: str, user_agent: str = "ai-agent-search-optimization/0.1", timeout: int = 12) -> FetchResult:
    req = urllib.request.Request(url, headers={"User-Agent": user_agent, "Accept": "text/html,text/plain,*/*"})
    try:
        with urllib.request.urlopen(req, timeout=timeout) as response:
            raw = response.read(2_000_000)
            charset = response.headers.get_content_charset() or "utf-8"
            return FetchResult(
                url=response.geturl(),
                status=response.status,
                headers=dict(response.headers.items()),
                text=raw.decode(charset, errors="replace"),
                error=None,
            )
    except urllib.error.HTTPError as exc:
        body = exc.read(200_000).decode("utf-8", errors="replace")
        return FetchResult(url=url, status=exc.code, headers=dict(exc.headers.items()), text=body, error=str(exc))
    except Exception as exc:  # noqa: BLE001
        return FetchResult(url=url, status=None, headers={}, text="", error=str(exc))


def meta_content(parser: PageParser, *names: str) -> str:
    wanted = {name.lower() for name in names}
    for item in parser.meta:
        key = (item.get("name") or item.get("property") or "").lower()
        if key in wanted:
            return item.get("content", "")
    return ""


def link_href(parser: PageParser, rel_name: str) -> str:
    rel_name = rel_name.lower()
    for item in parser.links:
        rels = {part.lower() for part in item.get("rel", "").split()}
        if rel_name in rels:
            return item.get("href", "")
    return ""


def parse_schema_types(json_ld_blocks: list[str]) -> list[str]:
    types: list[str] = []

    def walk(value: Any) -> None:
        if isinstance(value, dict):
            kind = value.get("@type")
            if isinstance(kind, str):
                types.append(kind)
            elif isinstance(kind, list):
                types.extend(str(item) for item in kind)
            for child in value.values():
                walk(child)
        elif isinstance(value, list):
            for child in value:
                walk(child)

    for block in json_ld_blocks:
        try:
            walk(json.loads(block))
        except json.JSONDecodeError:
            continue
    return sorted(set(types))


def parse_sitemaps(robots_text: str, site_origin: str) -> list[str]:
    found = []
    for line in robots_text.splitlines():
        if line.lower().startswith("sitemap:"):
            found.append(line.split(":", 1)[1].strip())
    if not found:
        found.append(urllib.parse.urljoin(site_origin, "/sitemap.xml"))
    return found


def audit(url: str, brand: str | None, agents: list[str]) -> dict[str, Any]:
    target = normalize_url(url)
    site_origin = origin(target)
    robots_url = urllib.parse.urljoin(site_origin, "/robots.txt")
    llms_url = urllib.parse.urljoin(site_origin, "/llms.txt")

    robots = fetch(robots_url, timeout=8)
    page = fetch(target)
    llms = fetch(llms_url, timeout=8)

    robot_parser = urllib.robotparser.RobotFileParser()
    robot_parser.set_url(robots_url)
    if robots.get("status") == 200:
        robot_parser.parse(robots.get("text", "").splitlines())

    parser = PageParser()
    if page.get("text"):
        parser.feed(page["text"])
        parser.close()

    visible_text = html.unescape(" ".join(parser.text_parts))
    visible_text = re.sub(r"\s+", " ", visible_text).strip()
    robots_meta = meta_content(parser, "robots", "googlebot")
    schema_types = parse_schema_types(parser.json_ld)
    sitemap_urls = parse_sitemaps(robots.get("text", ""), site_origin) if robots.get("status") == 200 else []

    agent_access = {}
    for agent in agents:
        if robots.get("status") == 200:
            agent_access[agent] = robot_parser.can_fetch(agent, target)
        else:
            agent_access[agent] = None

    internal_links = []
    for link in parser.links:
        href = link.get("href", "")
        if not href:
            continue
        absolute = urllib.parse.urljoin(target, href)
        if urllib.parse.urlparse(absolute).netloc == urllib.parse.urlparse(site_origin).netloc:
            internal_links.append(absolute)

    findings = build_findings(
        page=page,
        robots=robots,
        llms=llms,
        parser=parser,
        visible_text=visible_text,
        robots_meta=robots_meta,
        schema_types=schema_types,
        agent_access=agent_access,
        sitemap_urls=sitemap_urls,
        brand=brand,
    )

    return {
        "url": target,
        "final_url": page.get("url"),
        "status": page.get("status"),
        "robots_url": robots_url,
        "llms_url": llms_url,
        "title": parser.title.strip(),
        "description": meta_content(parser, "description"),
        "canonical": link_href(parser, "canonical"),
        "robots_meta": robots_meta,
        "h1": parser.headings["h1"][:5],
        "h2_sample": parser.headings["h2"][:10],
        "visible_text_chars": len(visible_text),
        "schema_types": schema_types,
        "json_ld_blocks": len(parser.json_ld),
        "internal_link_count": len(set(internal_links)),
        "sitemaps": sitemap_urls,
        "crawler_access": agent_access,
        "llms_txt": {
            "status": llms.get("status"),
            "chars": len(llms.get("text", "")),
            "has_h1": bool(re.search(r"^#\s+", llms.get("text", ""), re.M)),
            "link_count": len(re.findall(r"\[[^\]]+\]\([^)]+\)", llms.get("text", ""))),
        },
        "findings": findings,
    }


def build_findings(**data: Any) -> list[dict[str, str]]:
    findings: list[dict[str, str]] = []

    def add(severity: str, title: str, detail: str, fix: str) -> None:
        findings.append({"severity": severity, "title": title, "detail": detail, "fix": fix})

    page = data["page"]
    robots = data["robots"]
    llms = data["llms"]
    parser = data["parser"]
    visible_text = data["visible_text"]
    robots_meta = data["robots_meta"].lower()
    schema_types = data["schema_types"]
    agent_access = data["agent_access"]
    sitemap_urls = data["sitemap_urls"]
    brand = data["brand"]

    if page.get("status") != 200:
        add("critical", "Target URL is not returning HTTP 200", f"Status: {page.get('status')}, error: {page.get('error')}", "Fix status, redirect chain, or target URL before optimizing content.")
    if robots.get("status") != 200:
        add("high", "robots.txt could not be fetched", f"Status: {robots.get('status')}, error: {robots.get('error')}", "Publish a valid robots.txt and include sitemap references.")
    for agent, allowed in agent_access.items():
        if allowed is False:
            add("high", f"{agent} is disallowed by robots.txt", "This can block AI/search retrieval for that crawler.", f"Review robots.txt policy for {agent}; allow search crawlers when visibility is desired.")
    if not sitemap_urls:
        add("medium", "No sitemap discovered", "No Sitemap directive was found in robots.txt.", "Add Sitemap directives to robots.txt and keep XML sitemaps current.")
    if "noindex" in robots_meta:
        add("critical", "Page has noindex directive", robots_meta, "Remove noindex from pages meant to appear in search or AI answers.")
    if "nosnippet" in robots_meta or "max-snippet:0" in robots_meta.replace(" ", ""):
        add("high", "Snippet restrictions may reduce AI feature eligibility", robots_meta, "Use snippet restrictions only where required by policy or licensing.")
    if len(visible_text) < 1000:
        add("medium", "Low visible text volume", f"Only {len(visible_text)} visible text characters parsed.", "Ensure key answers, product facts, and evidence are available as server-rendered or crawlable text.")
    if not parser.title.strip():
        add("medium", "Missing title tag", "No title was parsed.", "Add a concise title naming the entity, category, and page purpose.")
    if not parser.headings["h1"]:
        add("medium", "Missing H1", "No H1 was parsed.", "Add one descriptive H1 that identifies the page topic.")
    if not schema_types:
        add("medium", "No JSON-LD schema detected", "No valid application/ld+json @type values found.", "Add schema that matches visible content, such as Organization, Product, Service, Article, FAQPage, or BreadcrumbList.")
    if llms.get("status") != 200:
        add("low", "No /llms.txt found", f"Status: {llms.get('status')}", "Consider adding a concise /llms.txt map for AI agents if the site has documentation, product, policy, or support resources.")
    if brand and brand.lower() not in visible_text.lower():
        add("medium", "Brand/entity name not found in visible text", f"Brand checked: {brand}", "Make the primary entity explicit in crawlable copy, schema, title, and key landing pages.")
    return findings


def to_markdown(report: dict[str, Any]) -> str:
    lines = [
        f"# AI Search Visibility Audit: {report['url']}",
        "",
        f"- Status: {report['status']}",
        f"- Final URL: {report.get('final_url')}",
        f"- Title: {report.get('title') or '(missing)'}",
        f"- Description: {report.get('description') or '(missing)'}",
        f"- Canonical: {report.get('canonical') or '(missing)'}",
        f"- Visible text chars: {report['visible_text_chars']}",
        f"- Schema types: {', '.join(report['schema_types']) or '(none)'}",
        f"- Internal links parsed: {report['internal_link_count']}",
        "",
        "## Crawler Access",
        "",
    ]
    for agent, allowed in report["crawler_access"].items():
        value = "unknown" if allowed is None else ("allowed" if allowed else "blocked")
        lines.append(f"- {agent}: {value}")
    lines.extend(["", "## llms.txt", ""])
    llms = report["llms_txt"]
    lines.extend([
        f"- URL: {report['llms_url']}",
        f"- Status: {llms['status']}",
        f"- Characters: {llms['chars']}",
        f"- Markdown H1: {llms['has_h1']}",
        f"- Markdown links: {llms['link_count']}",
        "",
        "## Findings",
        "",
    ])
    if not report["findings"]:
        lines.append("No major issues found by the lightweight audit. Run a deeper crawl and prompt visibility test next.")
    for item in report["findings"]:
        lines.extend([
            f"### [{item['severity']}] {item['title']}",
            "",
            f"- Detail: {item['detail']}",
            f"- Fix: {item['fix']}",
            "",
        ])
    return "\n".join(lines).strip() + "\n"


def main() -> int:
    parser = argparse.ArgumentParser(description="Audit a URL for AI search visibility basics.")
    parser.add_argument("url", help="URL to audit")
    parser.add_argument("--brand", help="Brand/entity name to check in visible text")
    parser.add_argument("--agents", help="Comma-separated user agents to test against robots.txt")
    parser.add_argument("--markdown", action="store_true", help="Print Markdown instead of JSON")
    args = parser.parse_args()

    agents = [item.strip() for item in args.agents.split(",")] if args.agents else DEFAULT_AGENTS
    report = audit(args.url, args.brand, [agent for agent in agents if agent])
    if args.markdown:
        sys.stdout.write(to_markdown(report))
    else:
        json.dump(report, sys.stdout, indent=2, ensure_ascii=False)
        sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/tools/audit-site.ts
import { type AuditReport, type Finding, asJsonResult, asTextResult, auditToMarkdown } from "../lib/format.js";
import { decodeHtml, fetchText, siteOrigin } from "../lib/web.js";

const DEFAULT_AGENTS = [
  "Googlebot",
  "OAI-SearchBot",
  "GPTBot",
  "ChatGPT-User",
  "PerplexityBot",
  "Perplexity-User",
  "ClaudeBot",
  "CCBot"
];

type AuditSiteInput = {
  url: string;
  brand?: string;
  agents?: string[];
  format?: "json" | "markdown";
};

type ParsedHtml = {
  title: string;
  description: string;
  canonical: string;
  robotsMeta: string;
  h1: string[];
  h2: string[];
  visibleText: string;
  jsonLdBlocks: string[];
  schemaTypes: string[];
  internalLinkCount: number;
};

export async function auditSite(input: AuditSiteInput) {
  const target = new URL(input.url.startsWith("http") ? input.url : `https://${input.url}`).toString();
  const origin = siteOrigin(target);
  const robotsUrl = `${origin}/robots.txt`;
  const llmsUrl = `${origin}/llms.txt`;
  const agents = input.agents?.length ? input.agents : DEFAULT_AGENTS;

  const [page, robots, llms] = await Promise.all([
    fetchText(target),
    fetchText(robotsUrl, { timeoutMs: 8_000, maxBytes: 500_000 }),
    fetchText(llmsUrl, { timeoutMs: 8_000, maxBytes: 500_000 })
  ]);

  const parsed = parseHtml(page.text, origin);
  const sitemaps = robots.status === 200 ? parseSitemaps(robots.text, origin) : [];
  const crawlerAccess = Object.fromEntries(
    agents.map((agent) => [agent, robots.status === 200 ? canFetch(robots.text, agent, new URL(target).pathname || "/") : null])
  ) as Record<string, boolean | null>;

  const report: AuditReport = {
    url: target,
    finalUrl: page.url,
    status: page.status,
    title: parsed.title,
    description: parsed.description,
    canonical: parsed.canonical,
    robotsMeta: parsed.robotsMeta,
    h1: parsed.h1.slice(0, 5),
    h2Sample: parsed.h2.slice(0, 10),
    visibleTextChars: parsed.visibleText.length,
    schemaTypes: parsed.schemaTypes,
    jsonLdBlocks: parsed.jsonLdBlocks.length,
    internalLinkCount: parsed.internalLinkCount,
    sitemaps,
    crawlerAccess,
    llmsTxt: {
      url: llmsUrl,
      status: llms.status,
      chars: llms.text.length,
      hasH1: /^#\s+/m.test(llms.text),
      linkCount: (llms.text.match(/\[[^\]]+\]\([^)]+\)/g) ?? []).length
    },
    findings: buildFindings({
      pageStatus: page.status,
      pageError: page.error,
      robotsStatus: robots.status,
      robotsError: robots.error,
      robotsMeta: parsed.robotsMeta,
      visibleText: parsed.visibleText,
      title: parsed.title,
      h1: parsed.h1,
      schemaTypes: parsed.schemaTypes,
      llmsStatus: llms.status,
      brand: input.brand,
      crawlerAccess,
      sitemaps
    })
  };

  if (input.format === "markdown") {
    return asTextResult(auditToMarkdown(report));
  }
  return asJsonResult(report);
}

function parseHtml(html: string, origin: string): ParsedHtml {
  const cleaned = html
    .replace(/<script\b(?![^>]*application\/ld\+json)[\s\S]*?<\/script>/gi, " ")
    .replace(/<style[\s\S]*?<\/style>/gi, " ")
    .replace(/<noscript[\s\S]*?<\/noscript>/gi, " ");

  const title = textOfFirst(html, /<title[^>]*>([\s\S]*?)<\/title>/i);
  const description = metaContent(html, "description");
  const robotsMeta = [metaContent(html, "robots"), metaContent(html, "googlebot")].filter(Boolean).join(", ");
  const canonical = attrFromFirst(html, /<link\b[^>]*rel=["'][^"']*\bcanonical\b[^"']*["'][^>]*>/i, "href");
  const h1 = texts(cleaned, /<h1\b[^>]*>([\s\S]*?)<\/h1>/gi);
  const h2 = texts(cleaned, /<h2\b[^>]*>([\s\S]*?)<\/h2>/gi);
  const jsonLdBlocks = texts(html, /<script\b[^>]*type=["']application\/ld\+json["'][^>]*>([\s\S]*?)<\/script>/gi, false);
  const schemaTypes = parseSchemaTypes(jsonLdBlocks);
  const visibleText = decodeHtml(
    cleaned
      .replace(/<[^>]+>/g, " ")
      .replace(/\s+/g, " ")
      .trim()
  );
  const internalLinkCount = new Set(
    Array.from(html.matchAll(/<a\b[^>]*href=["']([^"']+)["'][^>]*>/gi))
      .map((match) => {
        try {
          return new URL(match[1], origin).toString();
        } catch {
          return "";
        }
      })
      .filter((href) => href && new URL(href).origin === origin)
  ).size;

  return { title, description, canonical, robotsMeta, h1, h2, visibleText, jsonLdBlocks, schemaTypes, internalLinkCount };
}

function textOfFirst(html: string, pattern: RegExp) {
  const match = html.match(pattern);
  return match ? stripTags(match[1]) : "";
}

function texts(html: string, pattern: RegExp, strip = true) {
  const values: string[] = [];
  for (const match of html.matchAll(pattern)) {
    values.push(strip ? stripTags(match[1]) : match[1].trim());
  }
  return values.filter(Boolean);
}

function stripTags(value: string) {
  return decodeHtml(value.replace(/<[^>]+>/g, " ").replace(/\s+/g, " ").trim());
}

function metaContent(html: string, name: string) {
  const escaped = name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const pattern = new RegExp(`<meta\\b(?=[^>]*(?:name|property)=["']${escaped}["'])[^>]*>`, "i");
  return attrFromFirst(html, pattern, "content");
}

function attrFromFirst(html: string, elementPattern: RegExp, attr: string) {
  const element = html.match(elementPattern)?.[0] ?? "";
  const pattern = new RegExp(`${attr}=["']([^"']+)["']`, "i");
  return decodeHtml(element.match(pattern)?.[1]?.trim() ?? "");
}

function parseSchemaTypes(blocks: string[]) {
  const types = new Set<string>();

  function walk(value: unknown): void {
    if (Array.isArray(value)) {
      value.forEach(walk);
      return;
    }
    if (!value || typeof value !== "object") {
      return;
    }
    const record = value as Record<string, unknown>;
    const type = record["@type"];
    if (typeof type === "string") {
      types.add(type);
    } else if (Array.isArray(type)) {
      type.forEach((item) => types.add(String(item)));
    }
    Object.values(record).forEach(walk);
  }

  for (const block of blocks) {
    try {
      walk(JSON.parse(block));
    } catch {
      // Ignore invalid JSON-LD, but keep the block count in the report.
    }
  }

  return [...types].sort();
}

function parseSitemaps(robots: string, origin: string) {
  const sitemaps = robots
    .split(/\r?\n/)
    .filter((line) => /^sitemap:/i.test(line))
    .map((line) => line.split(/:(.*)/s)[1]?.trim())
    .filter(Boolean) as string[];
  return sitemaps.length ? sitemaps : [`${origin}/sitemap.xml`];
}

function canFetch(robots: string, agent: string, path: string) {
  const groups = parseRobotsGroups(robots);
  const lowerAgent = agent.toLowerCase();
  const matching = groups.filter((group) => group.agents.some((item) => item === "*" || item === lowerAgent));
  if (matching.length === 0) {
    return true;
  }

  let best: { directive: "allow" | "disallow"; length: number } | null = null;
  for (const group of matching) {
    for (const rule of group.rules) {
      if (!rule.path || !path.startsWith(rule.path)) {
        continue;
      }
      if (!best || rule.path.length > best.length || (rule.path.length === best.length && rule.directive === "allow")) {
        best = { directive: rule.directive, length: rule.path.length };
      }
    }
  }
  return best ? best.directive === "allow" : true;
}

function parseRobotsGroups(robots: string) {
  const groups: Array<{ agents: string[]; rules: Array<{ directive: "allow" | "disallow"; path: string }> }> = [];
  let current: { agents: string[]; rules: Array<{ directive: "allow" | "disallow"; path: string }> } | null = null;

  for (const rawLine of robots.split(/\r?\n/)) {
    const line = rawLine.replace(/#.*/, "").trim();
    if (!line.includes(":")) {
      continue;
    }
    const [rawKey, ...rest] = line.split(":");
    const key = rawKey.trim().toLowerCase();
    const value = rest.join(":").trim();
    if (key === "user-agent") {
      if (!current || current.rules.length > 0) {
        current = { agents: [], rules: [] };
        groups.push(current);
      }
      current.agents.push(value.toLowerCase());
    } else if ((key === "allow" || key === "disallow") && current) {
      current.rules.push({ directive: key, path: value });
    }
  }

  return groups;
}

function buildFindings(input: {
  pageStatus: number | null;
  pageError: string | null;
  robotsStatus: number | null;
  robotsError: string | null;
  robotsMeta: string;
  visibleText: string;
  title: string;
  h1: string[];
  schemaTypes: string[];
  llmsStatus: number | null;
  brand?: string;
  crawlerAccess: Record<string, boolean | null>;
  sitemaps: string[];
}) {
  const findings: Finding[] = [];
  const add = (severity: Finding["severity"], title: string, detail: string, fix: string) => {
    findings.push({ severity, title, detail, fix });
  };

  if (input.pageStatus !== 200) {
    add("critical", "Target URL is not returning HTTP 200", `Status: ${input.pageStatus ?? "unknown"}, error: ${input.pageError ?? "none"}`, "Fix status, redirect chain, or target URL before optimizing content.");
  }
  if (input.robotsStatus !== 200) {
    add("high", "robots.txt could not be fetched", `Status: ${input.robotsStatus ?? "unknown"}, error: ${input.robotsError ?? "none"}`, "Publish a valid robots.txt and include sitemap references.");
  }
  for (const [agent, allowed] of Object.entries(input.crawlerAccess)) {
    if (allowed === false) {
      add("high", `${agent} is disallowed by robots.txt`, "This can block AI/search retrieval for that crawler.", `Review robots.txt policy for ${agent}; allow search crawlers when visibility is desired.`);
    }
  }
  if (input.sitemaps.length === 0) {
    add("medium", "No sitemap discovered", "No Sitemap directive was found in robots.txt.", "Add Sitemap directives to robots.txt and keep XML sitemaps current.");
  }
  const robotsMeta = input.robotsMeta.toLowerCase();
  if (robotsMeta.includes("noindex")) {
    add("critical", "Page has noindex directive", input.robotsMeta, "Remove noindex from pages meant to appear in search or AI answers.");
  }
  if (robotsMeta.includes("nosnippet") || robotsMeta.replace(/\s/g, "").includes("max-snippet:0")) {
    add("high", "Snippet restrictions may reduce AI feature eligibility", input.robotsMeta, "Use snippet restrictions only where required by policy or licensing.");
  }
  if (input.visibleText.length < 1000) {
    add("medium", "Low visible text volume", `Only ${input.visibleText.length} visible text characters parsed.`, "Ensure key answers, product facts, and evidence are available as server-rendered or crawlable text.");
  }
  if (!input.title) {
    add("medium", "Missing title tag", "No title was parsed.", "Add a concise title naming the entity, category, and page purpose.");
  }
  if (input.h1.length === 0) {
    add("medium", "Missing H1", "No H1 was parsed.", "Add one descriptive H1 that identifies the page topic.");
  }
  if (input.schemaTypes.length === 0) {
    add("medium", "No JSON-LD schema detected", "No valid application/ld+json @type values found.", "Add schema that matches visible content, such as Organization, Product, Service, Article, FAQPage, or BreadcrumbList.");
  }
  if (input.llmsStatus !== 200) {
    add("low", "No /llms.txt found", `Status: ${input.llmsStatus ?? "unknown"}`, "Consider adding a concise /llms.txt map for AI agents if the site has documentation, product, policy, or support resources.");
  }
  if (input.brand && !input.visibleText.toLowerCase().includes(input.brand.toLowerCase())) {
    add("medium", "Brand/entity name not found in visible text", `Brand checked: ${input.brand}`, "Make the primary entity explicit in crawlable copy, schema, title, and key landing pages.");
  }

  return findings;
}
